docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+  - Best effort pods will not be evicted because of CPU constraints, they will simply be starved if CPU resources are not available.
+- Non-compressible resources
+  - For now, we are only implementing memory limits.
+  - We will always launch a best-effort pod, regardless of its memory requirements.
+  - Best effort pods will be evicted if a top tier pod needing the memory requirements comes in, or if some other top-tier pod asks for memory.
+  - Best-effort pods will not be allowed to exceed their memory limit.
+  - Extension: Consider distinguishing between best-effort pods that specify memory limits and give them even lower priority. Otherwise a best-effort pod without a memory limit could use up system resources and kill off other best-effort pods.
+- Extension: support other resources like network, disk.
+
+Resource guarantees for top tier pods
+- Will not change.
+- Extension: Consider giving top-tier pods that do not specify memory limits a lower memory priority than top tier pods that specify memory limits (but a higher priority than best-effort pods)
+
+## Policy
+
+The QoS design does not make assumptions about the “policy” used to label pods as best-effort and top-tier.
+
+In the first iteration, we will stick to a simple policy. Pods that specify CPU and memory limits for every container will be classified as top-tier, all other pods will be classified as best-effort. This policy offers protection to pods that specify resource requirements (from pods that have no limits and can go overboard in resource consumption), which gives them better resource guarantees. 
+
+Extension: In the future, we might add a field to PodSpec in the API, to allow users to specify whether a pod is best-effort or top-tier.
+
+## Alternatives Considered
+
+Designing a simple but flexible quality of service framework is very challenging because there are many possible applications that can use tiered pods in different ways. Here are a couple of alternatives considered, and reasons for why we are not implementing them.
+
+One alternative is that we put all best-effort pods into a parent cgroup, and restrict the memory of the parent cgroup based on resource limits specified by the top-tier pods. This alternative would reduce the chance of a system OOM kill because it prevents jobs from using too much memory, but it reduces utilization. Also, the alternative would require require constant adjustment of the memory limit of cgroups (using a user-space control loop), which is tricky, racy, and error-prone. In the proposal outlined in this design document, best-effort pods will be allowed to "eat" into top-tier pods' memory, but will be killed if top-tier pods needs the memory they reserved. This leads to better utilization. It is a difficult tradeoff, but the better utilization of the proposal is why we rejected this alternative.

```

Comment:
Memory expands like a gas. Do we need a user-space control loop regardless to apply memory pressure?